pkg/server/adminHandler.go

```diff
@@ -646,3 +648,41 @@ func (h *adminHandler) globalOptionsSet(w http.ResponseWriter, r *http.Request)
 	}
 	w.WriteHeader(http.StatusOK)
 }
+
+func (h *adminHandler) signingCertSet(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("bad body: %v", err), http.StatusBadRequest)
+		return
+	}
+	if len(body) == 0 {
+		http.Error(w, "empty body", http.StatusBadRequest)
+		return
+	}
+
+	// validate the cert
+	_, err = ax.ParseCert(body)
+	if err != nil {
+		log.Printf("signingCertSet: %s", err)
+		http.Error(w, fmt.Sprintf("failed to parse signing cert: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	certPath := *h.signingCertPath
+	// when no signing cert is set, we will create a new file
+	if certPath == "" {
+		// generate a random name
+		certPath = fmt.Sprintf("/tmp/signingCert-%s.pem", common.RandomString(8))
+	}
+	// write the cert to the file atomically (using a temp file)
+	err = util.WriteRename(certPath, body)
+	if err != nil {
+		log.Printf("signingCertSet: %s", err)
+		http.Error(w, fmt.Sprintf("failed to write signing cert: %s", err), http.StatusInternalServerError)
+		return
+	}
+	// set the signing cert path
+	*h.signingCertPath = certPath
```

Comment:
As I understand we try to override the variable which come from [cli](https://github.com/lf-edge/adam/blob/9f6fca2a03ff210a38f037680db31c8e47c70496/cmd/server.go#L265) here. I can see that we potentially may create new file few lines above, how we will use it in case of adam's conatiner restart?
I still cannot understand, is the approach you define here better than just manual file [override](https://github.com/lf-edge/adam/issues/118#issuecomment-1745540885). Looks like restart of adam is not required to re-read the certificate.

Reply:
We only create a different file in case that CLI option was set to an empty string, otherwise we rewrite the certificate file. I think it's pretty unlikely that somebody unsets the CLI option explicetly, if it's not set we use the default value. So if Adam is restarted with the same CLI params it will pick up the new certificate. Only if it's restarted with `--signing-cert=""` it will not pick up any certificate and we'll need to set it again with the POST request.

I didn't understand how you are proposing to replace the certificate file from Eden without an API call? I thought they are only supposed to communicate through the API and we cannot assume that Adam and Eden are running on the same system, so that we can just replace the file locally.

Reply:
>  it's pretty unlikely that somebody unsets the CLI option explicitly

well, sounds reasonable, it will look like a problem of the caller.

> I didn't understand how you are proposing to replace the certificate file from Eden without an API call? I thought they are only supposed to communicate through the API and we cannot assume that Adam and Eden are running on the same system, so that we can just replace the file locally.

right now we prepare the options to start Adam container in Eden. And we generate certs on Eden side. But looks like we use environment variables to set signing cert [here](https://github.com/lf-edge/eden/blob/c2ba0752c0587dae3f5dd5d2fa74f56f6fcf2e8c/pkg/eden/eden.go#L158). So we cannot change the certificate without Adam's restart (`StopAdam` + `StartAdam` calls) in current logic. Can you please check that we will have no problems without file replacement after Adam restart (I mean, will fill correct `SIGNING_CERT` env)?

Reply:
upon the start of adam server, we always take the certificate from the env - so the one that's provided by the caller. So now it looks like it's eden's responsibility to provide the new certificate / replace the certificate file etc.

Reply:
btw, why is the `adam stop` command doing the exact [opposite](https://github.com/lf-edge/eden/blob/a3900ce0d6317a6188d3c2625379ff4fcfbf9ece/pkg/eden/eden.go#L231-L240) of what `--adam-rm` requests? and all the other stop commands as well?

Reply:
> btw, why is the `adam stop` command doing the exact [opposite](https://github.com/lf-edge/eden/blob/a3900ce0d6317a6188d3c2625379ff4fcfbf9ece/pkg/eden/eden.go#L231-L240) of what `--adam-rm` requests? and all the other stop commands as well?

Good question, looks like a bug.

Reply:
right now it works like this:
- the cli parameter `--signing-cert` is only providing the path to the certificate inside the container (imo basically useless)
- the signing certificate itself (it's content) is ALWAYS read by adam at the start from the env `SIGNING_CERT`. that means that the certificate (and adam's state in this regard) is effectively stored outside of adam by whoever provides the env (in our case it's Eden)

this poses a problem like @giggsoff and @deitch described that we don't preserve the state of the signing certificate when restarting adam's container - so I propose a simple check at the start of adam:
- if the file for the signing certificate already exists - just use it
- if not - take the content of the env `SIGNING_CERT`, put it in that file and use it (like we do now)

Reply:
Yeah. `signing-cert`, sorry for the incorrect flag for the right idea, and thanks for correcting me.

Reply:
Reading the code:

1. if env `SIGNING_CERT` is provided, it uses the contents of the env and writes it to the file specified by `--signing-cert` (this definitely is not how most people expect matching env var and flag to work); else
1. if `--autoCert`, create a cert and write it to the file specified by `--signing-cert` (which means that if you run it twice, it overrides)
1. load it from the file specified by `--signing-cert`

> we don't preserve the state of the signing certificate when restarting adam's container

Or adam as a process, for that matter. We are fine with losing state of adam when

* restart the process, if memory store
* restart a container, if file store and we didn't mount in from a persistent path

Coming back to the purpose of this PR, I think we are after updating a signing cert via API, so we can change it without restarting adam.

A few questions:

1. What is bad about restarting adam with a different cert? If there is other data we lose during restart, _even if in disk storage mode_, then we should fix that, then move on to helpful. Almost all app/web servers require a restart to change certs, rather than store in a database: nginx, apache, API servers, etc. etc. They just point to files on disk. Since the signing-cert is adam-wide (not per device), is it wrong to change a file on disk and restart? I get this may be easier, but what is _bad_ about restart?
2. Can we be really clear about which are _writable_ and which are _readable_? If adam will generate things and write to a path, then that path should not ever be populated by a user and read by adam on startup. It almost is guaranteed to create problems, "why did this overwrite my file??" E.g. is `--signing-cert` intended to say, "this is where adam will write a cert inside a database", in which case we should _never_ expose it? Or is it intended to say, "this is where adam will read a cert from, but it will never modify that file"? It should not be both.

Here is what is coming together in my head. I may be far off, so correct me.

* CLI flags and env vars parallel precisely, same behaviour (preferably, with `ADAM_` prefix for env vars); flags always override env vars (widely accepted behaviour, but beyond scope of this PR)
* flags and env vars are either to _override_ what is in the DB, or to _seed_ what will be in the DB, depends on flag. So if I put an env var `SIGNING_CERT_PATH` or cli flag `--server-cert-path`, that is where adam loads it from _on init only_. It then stores it in its db in a path that is opaque to the user. Whether that _overrides_ what is in the DB, or only fills it if the DB does not have one yet is TBD. Same for server-cert and everything else. Which means that there is _no_ default to `--server-cert-path`, because it is used to read only on startup.
* No flags/env vars let you change what is _inside_ the database.  Everything in there is opaque to the CLI. You can provide paths for _seeding_ or _override_ that are outside the database structure, in which case they are read-only to adam.
* Assuming we actually need an API and restart is not enough, using the API for `/certs` writes to the DB (in a location that the user has no control over or visibility to).

Yes, you can point out that some of these date back to when we first created adam and I had a hand in a lot of them. At least I am complaining about my own stuff publicly, and not someone else's! 😄

Reply:
@deitch you are proposing to overhaul the whole configuration system of Adam, which I am in favor of and agree with most of what you said! I would only propose not to use envs altogether and resort to only cli params, e.g. just giving adam the file with the signing certificate by putting it's path into `--signing-cert`. But this is out-of-scope for this PR so we should create a new issue for that.

> What is bad about restarting adam with a different cert? If there is other data we lose during restart, even if in disk storage mode, then we should fix that, then move on to helpful. Almost all app/web servers require a restart to change certs, rather than store in a database: nginx, apache, API servers, etc. etc. They just point to files on disk. Since the signing-cert is adam-wide (not per device), is it wrong to change a file on disk and restart? I get this may be easier, but what is bad about restart?
#118 has a discussion about that. Imo a restart would be the easiest way to do the certificate change, but we are going to lose the DB data in the case of in-memory driver. Also we should probably differentiate losing the info in the DB (client data) and losing the server configuration such as certificates (server data or metadata)

Reply:
Overhaul? Nah, just "make improvements which may be more than a little incremental" 😂

Reply:
In any case, you are right. What I'm raising is much wider than this, and your PR shouldn't be stuck because of it.

I raise it more because as we touch some parts, we at least should move in the right direction, which means we need agreement on what "right direction" is.

Yes, restart with memory driver loses whatever is in memory. That's why it's a memory driver. If we need to change the cert and not lose the edge device info, isn't it simpler to just run with file driver?

Reply:
I'd like to redirect this question to @milan-zededa @giggsoff @eriknordmark  :)

Reply:
I am still voting for restart with file/env change.
Potentially we should double check that during teardown (sigterm) we wait for handlers to write all changes into redis/files and avoid new handlers to receive the data.

Reply:
SIGHUP. That's interesting and quite common that SIGHUP is used to say "reload config". Having adam accept SIGHUP (kill -1) means "keep everything in memory but reload on disk config and cert/key files" is a nice elegant approach.

Reply:
Actually I wrote about graceful shutdown of container. But you point is valid.

Reply:
@deitch @giggsoff What are your objections against a unified Adam API for all possible run-time config changes (incl. certificate change)?

Reply:
Don't object to it per se, as long as it has clear expected behaviour. Meaning that if we pass it an updated cert via API, what happens to the one on disk (should not overwrite it)? What happens when we restart with the exact same commands (e.g. container process dies so automatically restarts)?

As long as the behaviour is clear and consistent, no issue at all.

The idea of SIGHUP (Petr's idea, give him credit) is that, most things that have some certs and configs as files, either restart to read new ones, or use a signal to indicate "reload". That might be easier than resolving the above.

Reply:
Certificate we are talking about here come from [CLI](https://github.com/lf-edge/adam/blob/9f6fca2a03ff210a38f037680db31c8e47c70496/cmd/server.go#L265) (path to cert) or [ENV](https://github.com/lf-edge/adam/blob/9f6fca2a03ff210a38f037680db31c8e47c70496/cmd/server.go#L93) (content of the certificate) options during startup of Adam. From Eden right now we [set](https://github.com/lf-edge/eden/blob/a3900ce0d6317a6188d3c2625379ff4fcfbf9ece/pkg/eden/eden.go#L158) the certificate (not the path, the cert itself) using ENV. How can we merge certificate come from ENV and certificate come through novel API? Looks like we will be forced to override ENV (and re-create container) in both cases (novel API and old+restart).

Again I have no good ideas about SIGHUP. I just want to handle `docker stop` with graceful shutdown of `gorilla/mux`.

Reply:
What about using the mounted file or the env. variable with the certificate only for the initial startup and then let adam move all its state (incl. certs) to redis (or whatever storage is used)? That way there will be only one source of truth - the database.
Just an idea. But in general I don't have strong opinion about this, so whatever will get this PR unblocked is fine by me.

Reply:
Guys, this discussion is clearly not working - if you want to discuss new approaches we should close this PR, open up a new issue and discuss it there, then come up with an approach that satisfies everyone and only then start implementing it.

My goal for now is just to decide how to proceed with this PR, if it's missing something and we can modify it and then merge it as quickly as possible - let's do it and discuss further. If not - let's do what I described above :arrow_up: 


Here is how it works now:
- Eden starts Adam and gives it the signing certificate through an env.
- If Adam doesn't already have a signing certificate in a certain path, it takes the contents of the env, writes them to the path and uses that as signing certificate
- Else - Adam just uses the signing certificate that it already has (consistency on Adam restart)
- Adam exposes a POST API through which we can set the signing certificate - in this case the payload of POST will overwrite Adam's signing certificate file. Since Eden starts Adam in a container - no changes outside the container are made by this operation.

Thus we have
1. Ability to "seed" signing certificate for the first start of Adam :heavy_check_mark: 
2. Ability to keep the signing certificate when Adam is restarted :heavy_check_mark: 
3. Ability to set the signing certificate during Adam's runtime, so that no restart is required and thus no data from the DB is lost :heavy_check_mark: 


Please tell me what the current solution is missing and we can work on that :pray:

Reply:
One quick follow-up: if you want to do this really quickly and easily, and avoid all of the mess:

1. Keep everything the way it is.
2. Create a signal handler for EVE to reload all config and certs (but not database) on SIGHUP (I think @giggsoff already did that)
3. If you want, add an API call to do the same thing (reload all certs and config, like SIGHUP)

I think that is the fastest way to get what you want, without getting bogged down in the mud that the current structure created for you.

I also think it is the right thing to do. Keep it on disk like apache and nginx and every other program that uses certs, just add the ability to reload them. Deal with the config issues another day.

Reply:
I just reread the EVE code. As far as I can tell, it reads the certs from disk with every single call to `GET /certs`. It doesn’t even cache them in memory, and so doesn't need SIGHUP or some "POST call to reload". Changing it on disk should change it on the next GET call.

[getAllCerts()](https://github.com/lf-edge/adam/blob/aad2368d41ed5842d314ede2bd82bab8637300ec/pkg/server/apiHandlerv2.go#L133) calls [getCertChain()](https://github.com/lf-edge/adam/blob/aad2368d41ed5842d314ede2bd82bab8637300ec/pkg/server/apiHandlerv2.go#L204), which just [reads it from disk](https://github.com/lf-edge/adam/blob/aad2368d41ed5842d314ede2bd82bab8637300ec/pkg/server/apiHandlerv2.go#L205-L212):

```go
	certData, err := ioutil.ReadFile(certPath)
	if err != nil {
		return nil, err
	}
```

If this is correct, and you already know the path, just update the path?

Reply:
> if the certificate resides in the database path (like done by default), we shouldn't probably just manipulate files there, circumventing Adam's interfaces

100%. This usage of `--cert-path` is just wrong. Either there should be _no_ `--cert-path` option, or if there is one, it is outside the database and is a read-only space that Adam uses to get the content, never to write it.

> we can use it in a scenario when the caller (like Eden) and Adam are not colocated - through the network

Don't you find it strange to change those certs over the network? No other server software of which I am aware does that.

If you do, you are saying, "cert is a database artifact". In that case:

* `--cert-path` must be removed; it is fixed in the database
* cert must be stored in the actual database, based on the driver (file, redis, memory, etc.), but entirely inside the database; Adam has no knowledge of what it is or where it goes, just passed it to the database driver or reads it from the database driver
* the env var must be resolved. Does it _override_ what is in the database? Or _seed_ it? Override seems to make sense, unless we rename it something that says very clearly, "this is the seed only, but if something is in the database I will use that". Which still is a bit odd behaviour.

Doesn't `GET /certs` return multiple certs? What would your proposed `POST` do?

Reply:
I was talking about the POST implemented as part for this PR, just for the signing certificate

Reply:
What does `GET /certs` get? Is that just the signing certificate as well?

Reply:
I added `POST /admin/certs`. There is no `GET /admin/certs`